### PR TITLE
add helper to lib/attrsets: hasAttrByPath

### PR DESCRIPTION
**Pull request by zagy**
Make it easy to test if an attrset has an attribute by path.

### CODE REVIEW

File: lib/attrsets.nix [hunk 1]
```diff
@@ -23,6 +23,17 @@ rec {
       then attrByPath (tail attrPath) default e.${attr}
       else default;
 
+  /* Return an if an attribute from nested attribute set exists.
```
Comment (edolstra) [inline lib/attrsets.nix]: "an if an"

Comment (zagy) [inline lib/attrsets.nix]: Thanks. Fixed.

**Comment by zagy** [comment]
ping?

File: lib/attrsets.nix [hunk 2]
```diff
@@ -23,6 +23,17 @@ rec {
...
 
+  /* Return if an attribute from nested attribute set exists.
+     For instance ["x" "y"] applied to some set e returns true, if e.x.y exists. False
+     is returned otherwise. */
+  hasAttrByPath = attrPath: e:
+    let attr = head attrPath;
+    in
+      if attrPath == [] then true
+      else if e ? ${attr}
+      then hasAttrByPath (tail attrPath) e.${attr}
```
Comment (edolstra) [inline lib/attrsets.nix]: Using `tail` causes this function to have O(n^2) complexity. Probably not a big deal...

Comment (zagy) [inline lib/attrsets.nix]: Given the expected lenght of attrPath it's probably irrelevant. But then: how could it be implemented differently? The head/tail pattern is quite common in functional programming. What's the alternative?

Comment (edolstra) [inline lib/attrsets.nix]: By passing an index that gets incremented on every recursive call. See e.g. `fold` in `lib/lists.nix`.

**Comment by copumpkin** [comment]
It seems unfortunate that `hasAttrByPath` can't be expressed cleanly in terms of `attrByPath`, yet has almost exactly the same shape of logic.

**Comment by MarcWeber** [comment]
(a.b or null) will be what you can use in many cases

File: lib/attrsets.nix (hunk 2 above)
Comment (copumpkin) [inline lib/attrsets.nix]: The whole thing could probably be done quite cleanly as an application of `fold`, too. Both this and `attrByPath`.

**Pull request merged**